src/machine.rs

```diff
@@ -1183,6 +1194,11 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for MiriMachine<'mir, 'tcx> {
         mut frame: Frame<'mir, 'tcx, Provenance, FrameData<'tcx>>,
         unwinding: bool,
     ) -> InterpResult<'tcx, StackPopJump> {
+        if frame.extra.is_user_relevant {
+            // All that we store is whether or not the frame we just removed is local, so now we
+            // have no idea what if any the next topmost local frame is.
+            ecx.active_thread_mut().set_top_user_relevant_frame(None);
```

Comment:
I guess the alternative would be to do a search here, right? Then we'd always have a hot cache, which simplifies logic elsewhere so it might be worth it.

In fact we could speed that up by storing a 'pointer' to the next user-relevant frame -- on push, remember the current user-relevant frame and store it in the new frame as `prev_user_relevant_frame`, on pop restore that. Both of these are O(1) operations.

Reply:
:facepalm: I guess I never tried this. Yes. This code path is so incredibly cold that it doesn't make sense to only invalidate here.